File: src/lib/ask.ts
import { invoke } from "@tauri-apps/api/core";
import type { AskMessage, AskThread, AskThreadDetail, EffortLevel } from "../types";

/**
 * Typed IPC wrappers for Ask — the commands in `src-tauri/src/commands/ask.rs`.
 *
 * Storage and lifecycle, plus sending a turn and the three events it streams
 * over (`application/ask/events.rs`).
 *
 * **Rejections are plain strings** — every command returns `Result<T, String>`,
 * the same convention `lib/discovery.ts` documents.
 */

/**
 * Mirrors `ask_create`. No worktree and no agent process yet — both wait for
 * the first turn that needs them.
 *
 * `machineId` is the picker's value, on the same terms as
 * {@link import("./discovery").createDiscovery}'s: a blank one is the same as
 * none, which takes the project's own host.
 *
 * `network` is the thread's opening web-access posture. It has to be settled
 * here rather than by a later {@link updateAskThreadSettings}, because the
 * first turn reads the stored value at send time and cannot be walked back
 * once it has run.
 */
export async function createAskThread(input: {
  projectId: string;
  title: string;
  agentKind: string;
  model: string | null;
  effort: EffortLevel | null;
  machineId: string | null;
  network: boolean;
}): Promise<AskThread> {
  return invoke<AskThread>("ask_create", {
    input: {
      project_id: input.projectId,
      title: input.title,
      agent_kind: input.agentKind,
      model: input.model,
      effort: input.effort,
      machine_id: input.machineId,
      network: input.network,
    },
  });
}

/** Mirrors `ask_list` — a project's Ask threads, most recently touched first. */
export async function listAskThreads(projectId: string): Promise<AskThread[]> {
  return invoke<AskThread[]>("ask_list", { projectId });
}

/** Mirrors `ask_load` — an Ask thread and its whole transcript. */
export async function loadAskThread(threadId: string): Promise<AskThreadDetail> {
  return invoke<AskThreadDetail>("ask_load", { threadId });
}

/**
 * Mirrors `ask_running` — whether a turn is under way on this thread right
 * now. Asked once when a thread is selected: `ask_turn_status` carries every
 * later change, and a surface that mounted mid-turn has already missed the
 * transition that would have told it.
 */
export async function askTurnRunning(threadId: string): Promise<boolean> {
  return invoke<boolean>("ask_running", { threadId });
}

/** Mirrors `ask_rename`. */
export async function renameAskThread(threadId: string, title: string): Promise<AskThread> {
  return invoke<AskThread>("ask_rename", { threadId, title });
}

/** Mirrors `ask_delete` — deletes the thread and its transcript, via the
 *  declared foreign key. */
export async function deleteAskThread(threadId: string): Promise<void> {
  return invoke<void>("ask_delete", { threadId });
}

/**
 * Mirrors `ask_update_settings`. A key absent from `patch` is left alone; a
 * key set to `null` (`model`, `effort`) clears it. `agent_kind` has no field
 * here — a thread's harness is fixed at creation.
 *
 * `null` is load-bearing rather than a spelling of "no change": it survives to
 * Rust only because `AskSettingsPatch` deserializes those two fields through a
 * `present` helper, since serde's derive collapses `null` and absent alike into
 * "leave alone". Pass a key only when you mean to write it.
 */
export async function updateAskThreadSettings(
  threadId: string,
  patch: { model?: string | null; effort?: EffortLevel | null; network?: boolean },
): Promise<AskThread> {
  return invoke<AskThread>("ask_update_settings", {
    threadId,
    patch: {
      ...("model" in patch ? { model: patch.model } : {}),
      ...("effort" in patch ? { effort: patch.effort } : {}),
      ...("network" in patch ? { network: patch.network } : {}),
    },
  });
}

/**
 * Mirrors `ask_send_turn`. Returns the user's own message, already
 * persisted — the assistant's answer arrives later over the events below.
 */
export async function sendAskTurn(threadId: string, text: string): Promise<AskMessage> {
  return invoke<AskMessage>("ask_send_turn", { threadId, text });
}

// ── The streaming contract (`application/ask/events.rs`) ───────────────────

/** Every `AgentEvent` of a turn, as it arrives. */
export const EVENT_ASK_AGENT_EVENT = "ask_agent_event";
/** A turn's phase: `setting_up`, then `running`, then `idle` or `error`. */
export const EVENT_ASK_TURN_STATUS = "ask_turn_status";
/** The completion signal — a multi-minute turn that ended silently would
 *  force the user to sit and watch it. */
export const EVENT_ASK_TURN_COMPLETED = "ask_turn_completed";

/**
 * Payload of `ask_agent_event`.
 *
 * `event` is the Rust `AgentEvent`, an enum serde-tagged by `kind` and wide
 * enough that mirroring the whole of it here would be a second copy to keep
 * in step for the sake of the variants a consuming surface reads. It stays
 * `unknown` behind guards there, the same convention `lib/discovery.ts` uses.
 */
export interface AskAgentEventPayload {
  thread_id: string;
  event: unknown;
}

/** Mirrors `AskTurnStatus`. */
export interface AskTurnStatusPayload {
  thread_id: string;
  status: string;
  reason: string | null;
}

/** Mirrors `AskTurnCompleted`. */
export interface AskTurnCompletedPayload {
  thread_id: string;
  title: string;
  message_id: string | null;
  ending: "success" | "interrupted" | "failed" | "environmental";
  reason: string | null;
  cost_usd: number;
  tokens: number;
  duration_ms: number;
}

File: src-tauri/src/commands/ask.rs
//! Tauri surface for Ask storage/lifecycle (`docs/PRD_DISCOVERY.md`-adjacent,
//! but not itself a Discovery — see [`demeteo_core::application::ask`]).
//!
//! Thin by construction: every decision below is made in
//! [`demeteo_core::application::ask`].

use crate::domain::ids::AskThreadId;
use crate::domain::models::{AskMessage, AskThread, EffortLevel};
use crate::state::AppContext;
use demeteo_core::application::ask::{self, turn, AskThreadDetail, NewAskThread, NodeResolution};
use demeteo_core::ports::ask::AskThreadPatch;
use serde::Deserialize;
use tauri::{Emitter, State};

#[tauri::command]
pub fn ask_create(ctx: State<'_, AppContext>, input: NewAskThread) -> Result<AskThread, String> {
    ask::create(&ctx, input)
}

#[tauri::command]
pub fn ask_list(ctx: State<'_, AppContext>, project_id: String) -> Result<Vec<AskThread>, String> {
    ask::list_for_project(&ctx, &project_id)
}

#[tauri::command]
pub fn ask_load(ctx: State<'_, AppContext>, thread_id: String) -> Result<AskThreadDetail, String> {
    ask::load(&ctx, &AskThreadId::from(thread_id))
}

/// Whether this thread has a turn running right now — what a surface that
/// mounted mid-turn asks once, on select, since `ask_turn_status` reports
/// transitions and never repeats one already made.
#[tauri::command]
pub fn ask_running(ctx: State<'_, AppContext>, thread_id: String) -> Result<bool, String> {
    Ok(ask::turn_running(&ctx, &AskThreadId::from(thread_id)))
}

#[tauri::command]
pub fn ask_rename(
    ctx: State<'_, AppContext>,
    thread_id: String,
    title: String,
) -> Result<AskThread, String> {
    ask::rename(&ctx, &AskThreadId::from(thread_id), &title)
}

#[tauri::command]
pub fn ask_delete(ctx: State<'_, AppContext>, thread_id: String) -> Result<(), String> {
    ask::delete(&ctx, &AskThreadId::from(thread_id))
}

/// Distinguishes a key that was present-and-`null` from one that was absent,
/// which serde's derive on `Option<Option<T>>` cannot: `Option::deserialize`
/// maps JSON `null` to `None`, so both spellings collapse to the same value
/// and every `is_some()` consumer downstream reads "leave alone".
///
/// Paired with `#[serde(default)]` — absence never reaches this function, so
/// the `Some` it always returns means exactly "the key was there".
fn present<'de, D, T>(deserializer: D) -> Result<Option<Option<T>>, D::Error>
where
    D: serde::Deserializer<'de>,
    T: Deserialize<'de>,
{
    Option::<T>::deserialize(deserializer).map(Some)
}

/// What a settings change may set. `agent_kind` is absent on purpose — a
/// thread's harness is fixed at creation, matching `AskThreadPatch`.
///
/// `model` and `effort` carry [`AskThreadPatch`]'s double `Option` all the way
/// out to JSON, which is the only place in `commands/` that happens — its
/// other nullable fields are built in Rust. Hence [`present`]; `network` is
/// non-nullable and takes the plain derive.
#[derive(Debug, Deserialize)]
pub struct AskSettingsPatch {
    #[serde(default, deserialize_with = "present")]
    pub model: Option<Option<String>>,
    #[serde(default, deserialize_with = "present")]
    pub effort: Option<Option<EffortLevel>>,
    #[serde(default)]
    pub network: Option<bool>,
}

#[tauri::command]
pub fn ask_update_settings(
    ctx: State<'_, AppContext>,
    thread_id: String,
    patch: AskSettingsPatch,
) -> Result<AskThread, String> {
    ask::update_settings(
        &ctx,
        &AskThreadId::from(thread_id),
        AskThreadPatch {
            model: patch.model,
            effort: patch.effort,
            network: patch.network,
            ..Default::default()
        },
    )
}

/// Send the user's turn and start the agent's, mirroring
/// `discovery_send_turn` (`commands::discovery`) exactly. The returned
/// message is the user's own, already persisted — the assistant's answer
/// arrives later over [`turn`]'s three events.
#[tauri::command]
pub async fn ask_send_turn(
    ctx: State<'_, AppContext>,
    app: tauri::AppHandle,
    thread_id: String,
    text: String,
) -> Result<AskMessage, String> {
    turn::send(
        &ctx,
        &AskThreadId::from(thread_id),
        &text,
        move |event, payload| {
            let _ = app.emit(event, payload);
        },
    )
    .await
}

#[tauri::command]
pub async fn ask_resolve_node(
    ctx: State<'_, AppContext>,
    thread_id: String,
    message_id: String,
    node_id: String,
) -> Result<NodeResolution, String> {
    ask::resolve_node(&ctx, &AskThreadId::from(thread_id), &message_id, &node_id).await
}

#[cfg(test)]
#[path = "../../tests/infrastructure/ask.rs"]
mod tests;
